File: sdk/nodejs-sdk/src/protocols/v0_6/WriteSchema.js
'use strict'
const utils = require('../../utils')
const Protocol = require('../Protocol')

/**
 * An interface for controlling a 0.6 WriteSchema protocol.
 * @extends Protocol
 */
class WriteSchema extends Protocol {
  /**
   * Constructor for the 0.6 WriteSchema object. This constructor creates an object that is ready to write a schema
   * the ledger.
   *
   * @param name The given name for the schema
   * @param version The given version of the schema
   * @param attrs A varargs list of attribute name in the schema
   * @param threadId the thread id of the already started protocol
   * @return 0.6 WriteSchema object
   *
   * @property {String} msgFamily - 'write'
   * @property {String} msgFamilyVersion - '0.6'
   * @property {String} msgQualifier - 'Community Qualifier'
   * @property {String} this.msgNames.WRITE_SCHEMA - 'write'
   */
  constructor (name, version, attrs, threadId = null) {
    const msgFamily = 'write-schema'
    const msgFamilyVersion = '0.6'
    const msgQualifier = utils.constants.EVERNYM_MSG_QUALIFIER
    super(msgFamily, msgFamilyVersion, msgQualifier, threadId)

    utils.DbcUtil.requireStringNotNullOrEmpty(name, 'name')
    utils.DbcUtil.requireStringNotNullOrEmpty(version, 'version')
    utils.DbcUtil.requireArrayNotContainNull(attrs, 'attrs')

    this.name = name
    this.version = version
    this.attrs = attrs

    /**
     Name for 'write' control message
     */
    this.msgNames.WRITE_SCHEMA = 'write'
    this.statuses = {}
    this.statuses.WRITE_SUCCESSFUL = 0
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #write
     */
  async writeMsg (context) {
    var msg = this._getBaseMessage(this.msgNames.WRITE_SCHEMA)
    msg = this._addThread(msg)
    msg.name = this.name
    msg.version = this.version
    msg.attrNames = this.attrs
    return msg
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #write
     */
  async writeMsgPacked (context) {
    return this.getMessageBytes(context, await this.writeMsg(context))
  }

  /**
     * Directs verity-application to write the specified Schema to the Ledger
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async write (context) {
    await this.sendMessage(context, await this.writeMsgPacked(context))
  }
}
module.exports = WriteSchema

File: sdk/python-sdk/verity_sdk/protocols/v0_6/WriteSchema.py
from verity_sdk.protocols.Protocol import Protocol
from verity_sdk.utils import EVERNYM_MSG_QUALIFIER
from verity_sdk.utils.DbcUtil import DbcUtil


class WriteSchema(Protocol):
    """
    The WriteCredentialDefinition protocol writes credential definitions to an Indy Identity Ledger
    (often the Sovrin Ledger). This protocol expect that the issuer has been setup.
    """

    MSG_FAMILY = 'write-schema'
    """the family name for the message family"""
    MSG_FAMILY_VERSION = '0.6'
    """the version for the message family"""

    # Messages
    WRITE_SCHEMA = 'write'
    """Name for 'write' control message"""

    STATUS = 'status-report'
    """Name for 'status-report' signal message"""

    def __init__(self, name: str, version: str, attrs: [str]):
        """
        Args:
            name (str): The given name for the schema
            version (str): The given version of the schema
            attrs (list[str]): A list of attribute name in the schema
        """
        super().__init__(
            self.MSG_FAMILY,
            self.MSG_FAMILY_VERSION,
            msg_qualifier=EVERNYM_MSG_QUALIFIER,
        )

        DbcUtil.require_string_not_null_or_empty(name, 'name')
        DbcUtil.require_string_not_null_or_empty(version, 'version')
        DbcUtil.require_array_not_contain_null(attrs, 'attrs')

        self.name = name
        self.version = version
        self.attrs = attrs

    async def write(self, context, endorser_did=None):
        """
        Directs verity-application to write the specified Schema to the Ledger

        Endorser did should be provided, only if override of default endorser is needed.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            endorser_did (str): Optional DID of an organization to endorse writing transactions to the ledger
        """
        await self.send_message(context, await self.write_msg_packed(context, endorser_did))

    def write_msg(self, context, endorser_did=None):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            endorser_did (str): Optional DID of an organization to endorse writing transactions to the ledger

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.WRITE_SCHEMA)
        msg['name'] = self.name
        msg['version'] = self.version
        msg['attrNames'] = self.attrs
        if endorser_did is not None:
            msg['endorserDID'] = endorser_did
        self._add_thread(msg)
        return msg

    async def write_msg_packed(self, context, endorser_did=None):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
            endorser_did (str): Optional DID of an organization to endorse writing transactions to the ledger

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.write_msg(context, endorser_did))
